File: 16-String-Part-II/python/01_z_function.py
class Solution:
    def search(self, pat: str, txt: str) -> List[int]:
        total = pat + "$" + txt
        n = len(total)
        z = [0] * n
        l, r = 0, 0

        for i in range(1, n):
            if i > r:
                l = r = i
                while r < n and total[r] == total[r - l]:
                    r += 1
                z[i] = r - l
                r -= 1
            else:
                idx = i - l
                if i + z[idx] <= r:
                    z[i] = z[idx]
                else:
                    l = i
                    while r < n and total[r] == total[r - l]:
                        r += 1
                    z[i] = r - l
                    r -= 1

        ans = []
        m = len(pat)
        for i in range(n):
            if z[i] == m:
                ans.append(i - m - 1)
        return ans

File: 16-String-Part-II/java/01-ZFunction.java
class Solution {
    ArrayList<Integer> search(String pat, String txt) {
        String total = pat + "$" + txt;
        int n = total.length();
        int[] z = new int[n];
        int l = 0, r = 0;

        for (int i = 1; i < n; i++) {
            if (i > r) {
                l = r = i;
                while (r < n && total.charAt(r) == total.charAt(r - l))
                    r++;
                z[i] = r - l;
                r--;
            } else {
                int idx = i - l;
                if (i + z[idx] <= r) {
                    z[i] = z[idx];
                } else {
                    l = i;
                    while (r < n && total.charAt(r) == total.charAt(r - l))
                        r++;
                    z[i] = r - l;
                    r--;
                }
            }
        }

        ArrayList<Integer> ans = new ArrayList<>();
        int m = pat.length();
        for (int i = 0; i < n; i++) {
            if (z[i] == m) {
                ans.add(i - m - 1);
            }
        }
        return ans;
    }
}
